pcp/PCPProcessList.c

```diff
@@ -86,29 +86,55 @@ void ProcessList_delete(ProcessList* pl) {
    free(this);
 }
 
-static inline unsigned long long PCPProcessList_adjustTime(unsigned long long t) {
-   return t / 10;
+static inline unsigned long Metric_instance_s32(int metric, int pid, int offset, unsigned long fallback) {
+   pmAtomValue value;
+   if (Metric_instance(metric, pid, offset, &value, PM_TYPE_32))
+      return value.l;
+   return fallback;
 }
 
-static void PCPProcessList_updateID(Process* process, int pid, int offset) {
+static inline unsigned long Metric_instance_u32(int metric, int pid, int offset, unsigned long fallback) {
    pmAtomValue value;
+   if (Metric_instance(metric, pid, offset, &value, PM_TYPE_U32))
+      return value.ul;
+   return fallback;
+}
 
-   if (Metric_instance(PCP_PROC_TGID, pid, offset, &value, PM_TYPE_U32))
-      process->tgid = value.ul;
-   else
-      process->tgid = 1;
+static inline unsigned long long Metric_instance_u64(int metric, int pid, int offset, unsigned long long fallback) {
+   pmAtomValue value;
+   if (Metric_instance(metric, pid, offset, &value, PM_TYPE_U64))
+      return value.ull;
+   return fallback;
+}
 
-   if (Metric_instance(PCP_PROC_PPID, pid, offset, &value, PM_TYPE_U32))
-      process->ppid = value.ul;
-   else
-      process->ppid = 1;
+static inline unsigned long long Metric_instance_time(int metric, int pid, int offset) {
+   pmAtomValue value;
+   if (Metric_instance(metric, pid, offset, &value, PM_TYPE_U64))
+      return value.ull / 10;
+   return 0;
+}
+
+static inline unsigned long long Metric_instance_ONE_K(int metric, int pid, int offset) {
+    pmAtomValue value;
+   if (Metric_instance(metric, pid, offset, &value, PM_TYPE_U64))
+      return value.ull / ONE_K;
+   return ULLONG_MAX;
+}
 
-   if (Metric_instance(PCP_PROC_STATE, pid, offset, &value, PM_TYPE_STRING)) {
-      process->state = value.cp[0];
+static inline char Metric_instance_char(int metric, int pid, int offset, unsigned char fallback) {
+   pmAtomValue value;
+   if (Metric_instance(metric, pid, offset, &value, PM_TYPE_STRING)) {
+      unsigned char uchar = value.cp[0];
       free(value.cp);
-   } else {
-      process->state = 'X';
+      return uchar;
    }
+   return fallback;
+}
+
+static void PCPProcessList_updateID(Process* process, int pid, int offset) {
+   process->tgid = Metric_instance_u32(PCP_PROC_TGID, pid, offset, 1);
+   process->ppid = Metric_instance_u32(PCP_PROC_PPID, pid, offset, 1);
+   process->state = Metric_instance_char(PCP_PROC_STATE, pid, offset, 'X');
```

Comment:
Is there some consistency for undefined states across other platforms? I've seen `'?'` being used on Darwin, though. Let's try not to introduce another such set … ;-)

Reply:
@BenBE heh, sure thing - '?' works too, will update shortly.